README.md
# NUS3BANK Editor
_by jam1garner (1.0-1.7 then support) and BlueDan aka BDS aka DanSyor (1.7-now)_

This is a GUI for Windows users to convert and import audio tracks into nus3bank files. Nus3bank files are basically a set of idsp audio tracks (for bgm, they most likely have only one track) that can feature a loop or not.

You need Python 2.7 installed to run it. Or since 2.20, Python 3.5! [Official Python website](https://www.python.org/)

Run `NUS3BANK Editor.pyw` either on your file explorer or on console.
Or since 2.10 you can use `NUS3BANK_Editor.bat`, that you can bind to be the default program to open files with nus3bank extension.


### Credits:
- Soneek who made tools essential for this
- everyone involved with vgmstream
- everyone involved with ffmpeg
- everyone who tested this tool and gave feedback

### Known issues:
* Crack noise at the end of loops
* You might have encoding problems (some error window popping up yelling at you something about UnicodeEncodeError) related to characters within the paths of files you're using. I try my best for you to avoid those but unfortunately the behaviour doesn't seem consistent from one computer to another, making it hard for me to clearly identify what goes wrong. If you're using Python 2.7 and encounter these issues, either be careful about any potentially exotic character in the paths of your files, or consider switching to Python 3.5.
    
### Troubleshooting
Be sure to grab the latest version, for now download links are posted on the [GBATemp Thread](http://gbatemp.net/threads/easy-nus3bank-editor-with-gui.426370/) by BlueDan as BDS.
Future releases will be available on [Github](https://github.com/DanSyor/NUS3BANK-Editor/releases). You can try downloading the zip for the master branch on the [main page for the Github project](https://github.com/DanSyor/NUS3BANK-Editor) between releases.

You may want to check the latest pages in case any specific hotfix was posted.

If you encounter any error, please provide (on the above thread or PM BlueDan on Discord) a clear and detailed explanation of what you're attempting to achive and how you do so. A screenshot of the error is highly appreciated.

You may want to use NUS3BANK_Editor_troubleshooting.bat to have a better chance to screenshot any error.
 * "When trying to use NUS3BANK_Editor.bat, an error pops up: ``Windows cannot find 'pythonw'. Make sure you typed the name correctly, and then try again.``
  * Looks like Python is not in your environment variable PATH. Follow the following instructions (these are for Python 2.7):
        - Run the Python installer, pick `Change Python` or `Modify`
        - At `Customize Python` assure you enable `Add python.exe to Path` and complete the process
        - It should now work. If it doesn't, go on.
        - Right click on the windows logo (or hit windows + pause and go to step 6)
        - Select system
        - Select advanced settings
        - Environment variables
        - Find path, click edit
        - Select new and enter `C:\Python27` (Note: this path may change depending on your installation or your version of Python; there should be a file called python.exe in that directory)
        - Select new (again) and enter `C:\Python27\Scripts` (Note: this path may change depending on your installation)
        - Select OK in all these windows
        - If you have any cmd opened, you should close them before checking your %path% again
        - Try first with user variables (which should be enough) then system variables
 * "I'd like to run the Python 3.5 version, how do I do that?"
  * You can uninstall Python 2 and install Python 3. Yet, you might wish to keep Python 2 installed and be able to chose which version of python you want to use.
    - Go to `C:\Python27` (or wherever your Python 2 installation is).
    - There sould be `python.exe` and `pythonw.exe`. Copy and and paste them into the same directory.
    - Rename these copies `python2.exe` and `pythonw2.exe` (respectively)
    - Go to `C:\Users\<your username>\AppData\Local\Programs\Python\Python35` (or wherever your Python 3 installation is). If you don't know where your installation is you can either:
        - do `echo %path%` in a cmd window (assuming Python 3 is in your path environment variable) and locate the part about Python 3
        - run the installer again, select `Modify` and check the directory
        - `System` > `Advanced settings` > `Environment Variables` > Look for `Path` > Look for Python 
    - There sould be `python.exe` and `pythonw.exe`. Copy and and paste them into the same directory.
    - Rename these copies `python3.exe` and `pythonw3.exe` (respectively)
    - With this setup you can chose to load Python 2 by typing `python2` or Python 3 by typing `python3` in a cmd window or within a batch file. This will also so called idle Python 2 (resp. Python 3) to be forced with `pythonw2` (resp. `pythonw3`). (idle is Python with no console: if you double click a file with py extension it will be run with a console, if you double click the same file by changing its extension by pyw, it will run with no console)
    - To change which version is the "default" one (`python` and `pythonw` for cmd window or bat files, probably not default program for opening .py or .pyw files): `System` > `Advanced settings` > `Environment Variables` > Look for `Path` (in user variables) it should have both versions, if it doesn't, remove `C:\Python27` entry from the global path and add it back to the user path. The default version will be which one is higher between `C:\Python27` and `C:\Users\<your username>\AppData\Local\Programs\Python\Python35`.
    - To change which version is run with `NUS3BANK_Editor.bat`, edit it and replace `pythonw` with either `pythonw2` or `pythonw3`

[Version history](HISTORY.md)


HISTORY.md
### Changelog:

_2.20_ - 2016/07/?
 * Fixed other encoding problems for input file
 * Fixed encoding problems regarding nus3bank extraction
 * Fully ported to Python 3.5: if you still have encoding problems, consider switching
 * Disabled log file creation and re-enable DOS output when it exists (= launched from python.exe, no pythonw.exe)
 * Doesn't ask anymore for a confirmation to close the file when no change has been done since last save
 * workspace/tmpspace folder now gets deleted when empty

_2.13b_ - 2016/06/04
 * Updated NUS3BANK_Editor.bat so no DOS console pops up
 * Fixed IOError 9 when trying to print to a non-existent DOS console
    
_2.13a_ - 2016/06/04
 * Fixed other encoding issues (paths related)
 * Made traceback visible for errors happening with nus3bank extraction

_2.13_ - 2016/06/03
 * Fixed playlist creation (on previous versions, an m3u could be left opened after nus3bank extraction failed, eventually causing an error when clearing workspace)
 * Changed colors for the path of the file (gray = no changes, no need to save; blue = changes (probably), should save)
 * Better algorithm for partial display of path
 * Alternates background color for tracks for better visibility
 * Various changes in how the app closes files/itself
 * Fixed encoding for errors (…)

_2.12_ - 2016/06/03
 * Hotfix so that it works either like in 2.10 or in 2.11 when deleting these temporary files (previous hotfix broke it for others)

_2.11_ - 2016/06/03
 * Hotfix for an issue when deleting temporary files without being in the main folder

_2.10_ - 2016/06/03
 * Fixed some menu entries with wrong action
 * Fixed default name when exporting
 * Opening a NUS3BANK now sets focus inside the list, so you can scroll directly
 * You can now preview a selection of tracks: either click the button, use the File menu, double click or hit enter/return
 * Right-clicking now allows you to select one track or can be used on a set of tracks already selected and grants the user with a context menu with replace/revert/preview/export options
 * Fixed something I forgot to edit within nus3bank.py and nus3inject.py
 * Better error handling (yes, again!)
 * Added a batch file that can be used to open and edit nus3bank files anywhere you can go with your file explorer (code edited so that everything works fine)
 * Window is now truly resizable
 * Added on-screen indicator of whether or not changes have apparently be applied or not (so that you save) (green = no changes since last save; red = maybe changes from last save)
 * It also gives you the path to the file you're currently editing (the .nus3bank extension is hidden)
 * A major design flow that could prevent two instances to work correctly simultaneously has been fixed
 * Fixed some encoding problems (no more UnicodeEncodingError?)

_2.02_ - 2016/06/01
 * Now supports directly DSP and HSP files (audio replacement)
 * Added the option to consider "unknown" extensions as VGM files (audio replacement)

_2.01_ - 2016/05/31
 * Fixed "Saving as..."
 * Opening nus3bank files and saving should now work for people who had an error trying to

_2.00_ - 2016/05/31
 * Now uses ffmpeg to support a LOT more file types (literally anything ffmpeg can convert to wav, from MP3 to video files)
 * Added an x-axis scrollbar
 * Entries go blue when replaced
 * Error handling enhanced (yes, again!)
 * Grouped some extensions together (like .wav with .ogg and .mp3; .bXstm together) + added wildcard extension for any extra extension that works (video files...) for easier use
 * Added the ability to revert replacements
 * Now uses a set workspace folder instead of a file-dependant folder
 * Saving when the user wants (injections into nus3bank are done only when actually saving instead of whenever replacing an idsp; replacing an idsp only converts the file to idsp and puts it in the preview playlist)
 * Ability to save as another nus3bank file added
 * Fixing playlist path, take 2
 * Reworked conversion to idsp for replacement sounds. Can loop audio on demand.
 * Menu entries with keyboard shortcuts have been added
 * The "Export to WAV" button has been removed and replaced with a menu entry that can also export an idsp with no conversion

_1.96_ - 2016/05/27
 * Now copies replacement audio files to tmp files to ensure filename to be supported

_1.95_ - 2016/05/27
 * Fixed a bug where replacement idsp was not properly located and probably no change happened
 * Added better error handling for the conversion of replacement audio files to IDSP
 * Now makes a backup of the nus3bank file when replacing its idsp's
 * Replacing idsp's in the file also replaces them within the playlist, backups are made for replaced idsp's
 * Fixing playlist path, take 1
 * Fixed a bug that could happen and prevent from opening idsp's or replacing them

_1.9_ - 2016/05/26
 * Fixed opening NUS3BANK files
 * Additional warnings
 * Fixed multiple file replacement
 * Added a confirmation window for file replacement

_1.8_ - 2016/05/25
 * Better error handling

_1.7_ - 2016/05/22
 * Added playlist support (and open in playlist option).
 * Also fixed a bug with certain nus3banks not properly opening.

_1.6_ - 2016/05/21
 * Added an "Export to WAV" option. Just select the IDSPs to export and hit the button and it will export to `/nus3bankeditor/<Name of *.nus3bank>/<name of *.idsp>.wav` this does support doing multiple at once.

_1.5_ - 2016/05/18
 * Added the ability to open idsps from within the tool through the default IDSP opener. To set this up install a something that can listen to idsps (example: foobar2000+vgmstream) and open an idsp in any way, when prompted select open with foobar (or whatever program you want)

_1.3_ - 2016/05/15
 * Added support for BRSTMs, BCSTMs, BFSTMs and OGGs.
 * Added more debugging related features.
 * Fixed bugs nobody cares about.

_1.1_ - 2016/05/14
 * Fixed some issues with renamed nus3bank's not properly opening
 * Added the ability to import idsp's

pyLib/xis.py
# coding: utf-8
from __future__ import absolute_import, division, print_function
import pyLib.six as six
# import six
import sys

''' Attempting to solve UnicodeEncodeError problems for Python 2
        and doing nothing for Python 3
'''

def encode(s):
    if six.PY3:
        return s
    else:
        return s.encode(sys.getfilesystemencoding()) # afaik this always seems to work

def decode(s):
    if six.PY3:
        return s
    else:
        try:
            print(u"Attempt 1: no decoding")
            return s + u''
        except UnicodeDecodeError or UnicodeEncodeError:
            try:
                print(u"Attempt 2: decoding from file system encoding")
                return s.decode(sys.getfilesystemencoding()) + u''
            except UnicodeDecodeError or UnicodeEncodeError:
                print(u"Attempt 3: decoding from console output encoding")
                return s.decode(sys.stdout.encoding) + u'' # weird, but maybe 1st solution doesn't always work?
        
# if len(sys.argv) > 1:
    # print(decode(sys.argv[1]))


pyLib/nus3inject.py
# coding: utf-8
from __future__ import absolute_import, division, print_function
import pyLib.six as six
import sys
from pyLib.util import *
import os
import struct
import string
import zlib

class strg(object):
    name = b""
    size = 0
    offset = 0
    def __init__(self, id):
        self.id = id

class nus3bank(object):
    contents = []
    propOffset = 0
    binfOffset = 0
    grpOffset = 0
    dtonOffset = 0
    toneOffset = 0
    junkOffset = 0
    packOffset = 0
    def __init__(self, size):
        self.size = size

class tone(object):
    name = b""
    ext = b""
    packOffset = 0
    size = 0
    data = None
    def __init__(self, offset, metaSize):
        self.offset = offset
        self.metaSize = metaSize

def inject(nusPath,replacePath,hexID):
    nus3 = open(nusPath, 'rb+')
    if nus3.read(4) != b"NUS3":
        nus3.seek(0)
        data = nus3.read()
        nus3_decom = zlib.decompress(data)
        nus3.close()
        nus3 = open(string.replace(nusPath, u'.nus3bank', u'_decompressed.nus3bank'), u'wb+')
        nus3.write(nus3_decom)
        nus3.seek(0)
        data = None
        nus3_decom = None
    else:
        nus3.seek(0)

    assert nus3.read(4) == b"NUS3"

    all = False    
    if hexID == u'all':
        toneID = 0xffff
        all = True
    else:
        toneID = int(hexID, 0)
    print((hex(toneID)))

    replacementSoundSize = os.path.getsize(replacePath)
    mainPackOffset = 0


    size = readu32le(nus3)
    #nus3.seek(-4,1)
    #nus3.write(struct.pack("<I", size + replacementSoundSize))
    #nus3.write(struct.pack("<I", 0))



    bank = nus3bank(size)
    assert nus3.read(8) == b"BANKTOC ", u'Not a bank archive!'
    tocSize = readu32le(nus3)
    contentCount = readu32le(nus3)
    offset = 0x14 + tocSize
    for i in range(contentCount):
        content = nus3.read(4)
        contentSize = readu32le(nus3)
        if content == b"PROP":
            propOffset = offset
            propSize = contentSize
        elif content == b"BINF":
            binfOffset = offset
            binfSize = contentSize
        elif content == b"GRP b":
            grpOffset = offset
            grpSize = contentSize
        elif content == b"DTON":
            dtonOffset = offset
            dtonSize = contentSize
        elif content == b"TONE":
            toneOffset = offset
            toneSize = contentSize
        elif content == b"JUNK":
            junkOffset = offset
            junkSize = contentSize
        elif content == b"MARK":
            markOffset = offset
            markSize = contentSize
        elif content == b"PACK":
            packOffset = offset
            packSize = contentSize
            packSizeOffset = nus3.tell() - 4
            #nus3.seek(-4, 1)
            #nus3.write(struct.pack("<I", packSize + replacementSoundSize))
        else:
            print(u'Unknown content type ' + six.text_type(content))
        offset += 8 + contentSize

    nus3.seek(binfOffset)
    assert nus3.read(4) == b"BINF"
    assert readu32le(nus3) == binfSize
    assert readu32le(nus3) == 0
    assert readu32le(nus3) == 3
    binfStringSize = readByte(nus3)
    binfString = nus3.read(binfStringSize-1)

    nus3.seek(1,1)
    #print binfString
    padding = (binfStringSize + 1) % 4
    #print padding
    if padding == 0:
        pass
    else:
        nus3.seek(abs(padding-4), 1)
    nus3ID = readu32le(nus3)
    #print hex(nus3ID)

    nus3.seek(toneOffset)
    assert nus3.read(4) == b"TONE"
    assert readu32le(nus3) == toneSize
    toneCount = readu32le(nus3)
    print(u"%s:%s:\"%s\"" % (nus3ID, hex(nus3ID), binfString))
    tones = []
    for i in range(toneCount):
        offset = readu32le(nus3) + toneOffset + 8
        metaSize = readu32le(nus3)
        tones.append(tone(offset, metaSize))
    
    for i in range(toneCount):
        if all == True:
            toneID = i
        if tones[i].metaSize <= 0xc:
            continue
        nus3.seek(tones[i].offset+6)
        tempByte = readByte(nus3)
        if tempByte > 9 or tempByte == 0:
            nus3.seek(5, 1)
        else:
            nus3.seek(1,1)
        stringSize = readByte(nus3)
    #    print hex(nus3.tell())
        tones[i].name = nus3.read(stringSize - 1)
        nus3.seek(1,1)
        #print b"\t" + hex(i) + b":" + tones[i].name
        padding = (stringSize + 1) % 4
        if padding == 0:
            nus3.seek(4, 1)
        else:
            nus3.seek(abs(padding-4) + 4, 1)
    #    assert readu32le(nus3) == 8
        nus3.seek(4,1)
        tones[i].packOffset = readu32le(nus3)
        tones[i].size = readu32le(nus3)
        if i == toneID:
            mainPackOffset = tones[i].packOffset
            origSoundSize = tones[i].size
            nus3.seek(-4, 1)
            tones[i].size = replacementSoundSize
            #nus3.write(struct.pack("<I", replacementSoundSize))
            #nus3.seek(packOffset + 8 + tones[i].packOffset + tones[i].size)
            #print(hex(nus3.tell()))
            #remainingPackData = nus3.read(packSize - tones[i].packOffset - tones[i].size)
            #nus3.seek(packOffset + 8 + tones[i].packOffset)
            #print(hex(nus3.tell()))
            replacementSound = open(replacePath, 'rb')
            tones[i].data = replacementSound.read()
            #nus3.write(replacementSound.read())
            #nus3.write(remainingPackData)
            #finalOffset = nus3.tell()
        else: 
            nus3.seek(packOffset + 8 + tones[i].packOffset)
            tones[i].data = nus3.read(tones[i].size)
        
    currentPackOffset = 0
    for i in range(toneCount):
        if tones[i].metaSize <= 0xc:
            continue
        nus3.seek(tones[i].offset+6)
        tempByte = readByte(nus3)
        if tempByte > 9 or tempByte == 0:
            nus3.seek(5, 1)
        else:
            nus3.seek(1,1)
        stringSize = readByte(nus3)
    #    print hex(nus3.tell())
        tones[i].name = nus3.read(stringSize - 1)
        nus3.seek(1,1)
        #print b"\t" + hex(i) + b":" + tones[i].name
        padding = (stringSize + 1) % 4
        if padding == 0:
            nus3.seek(4, 1)
        else:
            nus3.seek(abs(padding-4) + 4, 1)
    #    assert readu32le(nus3) == 8
        nus3.seek(4,1)
        tones[i].packOffset = currentPackOffset
        nus3.write(struct.pack(b"<I", currentPackOffset))
        nus3.write(struct.pack(b"<I", tones[i].size))
        nus3.seek(packOffset + 8 + tones[i].packOffset)
        nus3.write(tones[i].data)
        tones[i].data = None
        currentPackOffset += tones[i].size
        #tones[i].packOffset = readu32le(nus3)
        #tones[i].size = readu32le(nus3)
        b"""
        if 0xffffffff > tones[i].packOffset > mainPackOffset:
            nus3.seek(-8, 1)
            print ("Tone %s at offset %s" % (hex(i), hex(nus3.tell())))
            nus3.write(struct.pack("<I", tones[i].packOffset + replacementSoundSize - origSoundSize))
        b"""
    finalOffset = nus3.tell()
    nus3.seek(packOffset + 4)
    #nus3.write(struct.pack("<I", packSize + replacementSoundSize - origSoundSize))    
    nus3.write(struct.pack(b"<I", currentPackOffset))    
    nus3.seek(packSizeOffset)    
    #nus3.write(struct.pack("<I", packSize + replacementSoundSize - origSoundSize))    
    nus3.write(struct.pack(b"<I", currentPackOffset))    
    nus3.seek(4)
    #nus3.write(struct.pack("<I", bank.size + replacementSoundSize - origSoundSize))    
    nus3.write(struct.pack(b"<I", bank.size + currentPackOffset - packSize))    

    replacementSound.close()    
    nus3.seek(0)
    finalNus3 = nus3.read(finalOffset)
    nus3.close()
    #newNus3 = open(string.replace(nusPath, b".nus3bank", b".edit_nus3bank"), b"wb")
    newNus3 = open(nusPath, u'wb')
    newNus3.write(finalNus3)
    newNus3.close()

pyLib/multilistbox.py
# coding: utf-8
from __future__ import absolute_import, division, print_function
import pyLib.six as six
import pyLib.xis as xis
if six.PY3:
    from tkinter import *
    import tkinter.font as tkFont
    import tkinter.ttk as ttk
else:
    from Tkinter import *
    import tkFont
    import ttk


# colors for listbox elements
# color_l_default = u'black'
# color_l_change  = u'blue'
# color_l_bg = [u'white',u'#eeeeff']

# evenrow = u'evenrow'
# oddrow  = u'oddrow'
# linetag = [evenrow,oddrow]

class MultiListbox(object):
    
    def __init__(self, root, columns, padx=0, pady= 0):
        self.root=root
        self.columns = columns
        self.tree = None
        self._setup_widgets(padx,pady)
        self._sel_start_iid = 0
        self._sel_last_iid = 0
        self._items = []
        self._context_menu_v = Menu(root,bd=0)
        self._double_click_action_v = lambda e: ()
        self.linetags = []
    
    def _setup_widgets(self,padx,pady):
        # create a treeview with dual scrollbars
        container = ttk.Frame(self.root)

        self.tree = ttk.Treeview(container, columns=self.columns, show=u"headings")
        vsb = ttk.Scrollbar(container, orient=VERTICAL  , command=self.tree.yview)
        hsb = ttk.Scrollbar(container, orient=HORIZONTAL, command=self.tree.xview)
        self.tree.configure(yscrollcommand=vsb.set, xscrollcommand=hsb.set)
        vsb.config(command=self.tree.yview)
        hsb.config(command=self.tree.xview)
        vsb.pack(side=RIGHT , fill=Y)
        hsb.pack(side=BOTTOM, fill=X)
        self.tree.pack(fill=BOTH, padx=padx, pady=pady, expand=1)
        container.pack(fill=BOTH, expand=1)
        
        # for i in range(len(linetag)):
            # self.tree.tag_configure(linetag[i],background=color_l_bg[i])

        self.tree.bind(u"<Button1-Motion>", self._selection)
        self.tree.bind(u"<Button-1>"      , self._selection_start)
        self.tree.bind(u"<Double-1>"      , self._double_click_action_f)
        self.tree.bind(u"<Button-3>"      , self._context_menu_f)
        # self.tree.bind('<Return>', self._print_sel)
        
        for col in self.columns:
            self.tree.heading(col, text=col, command=lambda _col=col: self._sort_column(_col, False))
            # adjust the column's width to the header string
            self.tree.column(col, width=tkFont.Font().measure(col))
    
    def tag_remove(self, tags, items='all'):
        if items == 'all':
            items = self.tree.get_children('')
        for item in items:
            old_tags = list(self.tree.item(item, tags=None))
            new_tags = []
            for tag in old_tags:
                if not tag in tags:
                    new_tags.append(tag)
            self.tree.item(item,tags=new_tags)
    
    def linetags_set(self, linetags=u'default'):
        self.tag_remove(self.linetags)
        if linetags != u'default':
            self.linetags = linetags
        if self.linetags:
            i = 0
            for item in self.tree.get_children(''):
                tags = list(self.tree.item(item,tags=None))
                tags.append(linetags[i%len(linetags)])
                i+=1
                self.tree.item(item,tags=tags)
                
    def tag_configure(self,tag,foreground='',background='',font='',image=''):
        self.tree.tag_configure(tag,foreground=foreground,background=background,font=font,image=image)
        
    def tag_set(self,items,tags):
        if items == 'all':
            items = self.tree.get_children('')
        if tags:
            for item in items:
                o_tags = list(self.tree.item(item,tags=None))
                for tag in tags:
                    o_tags.append(tag)
                self.tree.item(item,tags=o_tags)
    
    def _context_menu_f(self,event):
        try:
            id = self.tree.identify_row(event.y)
        except IndexError:
            return
        if id == '':
            return
        if not id in self.tree.selection():
            self.selectAll()
            self.invertSelection()
            self.tree.selection_add(id)
        self._context_menu_v.post(event.x_root,event.y_root)
        
    def set_context_menu(self, menu):
        self._context_menu_v = menu
    
    def _double_click_action_f(self,event):
        if self.on_item(event):
            self._double_click_action_v(event)
    
    def set_double_click_action(self, action):
        self._double_click_action_v = action
    
    def insert(self, index, item):
        if self.linetags:
            iid = self.tree.insert('',index,values=item,tags=[self.linetags[len(self.tree.get_children(''))%len(self.linetags)]])
        else:
            iid = self.tree.insert('',index,values=item)
        self._items.append(iid)
        for i,v in enumerate(item):
            vWidth = tkFont.Font().measure(v)
            # print(i)
            # print (str(v)+' '+str(vWidth) +' | Col '+str(self.tree.column(self.columns[i],width=None)))
            if self.tree.column(self.columns[i],width=None) < vWidth:
                self.tree.column(self.columns[i],width=vWidth)
        # print(type(self.tree.item(iid,tags=None)))
        return iid
    
    def on_item(self,event):
        try:
            id = self.tree.identify_row(event.y)
        except IndexError:
            return ''
        # if id == '':
            # return ''
        return id
        
    def _sort_column(self, col, reverse):
        l = [(self.tree.set(k, col), k) for k in self.tree.get_children('')]
        # print (l)
        try:
            l.sort(key=lambda t: int(t[0]), reverse=reverse)
            # print(u'ints!')
        except:
            try:
                l.sort(key=lambda t: int(t[0],16), reverse=reverse)
                # print('hex!')
            except:
                l.sort(reverse=reverse)
                # print(u'strings...')

        for index, (val, k) in enumerate(l):
            self.tree.move(k, '', index)

        # reverse sort order for next sort
        self.tree.heading(col, command=lambda _col=col: self._sort_column(_col, not reverse))
        # reset sort order in other columns
        for col2 in self.columns:
            if col != col2:
                self.tree.heading(col2, command=lambda _col=col2: self._sort_column(_col, False))
        
        # reset row tags
        self.linetags_set()
                
    def _selection_start(self,event):
        self._sel_start_iid = self.on_item(event)
        self._sel_last_iid  = self._sel_start_iid
        
    def _selection(self, event):
        if self._sel_start_iid == '':
            return
        id = self.on_item(event)
        if id == '':
            return
        # print('ID: '+id)
        # print('Position: '+str(l.index(id)))
        if id == self._sel_last_iid:
            return
        l = self.tree.get_children('')
        id_cur = l.index(id)
        id_pre = l.index(self._sel_last_iid)
        id_ini = l.index(self._sel_start_iid)
        if id_cur > id_ini:
            if id_cur > id_pre:
                for x in range(id_ini,id_cur+1):
                    self.tree.selection_add(l[x])
            else:
                for x in range(id_cur+1,id_pre+1):
                    self.tree.selection_remove(l[x])
        elif id_cur < id_ini:
            if id_cur < id_pre:
                for x in range(id_cur,id_ini+1):
                    self.tree.selection_add(l[x])
            else:
                for x in range(id_pre,id_cur):
                    self.tree.selection_remove(l[x])
        else: # id = _sel_start_iid
            if id_cur < id_pre:
                for x in range(id_cur+1,id_pre+1):
                    self.tree.selection_remove(l[x])
            else:
                for x in range(id_pre,id_cur):
                    self.tree.selection_remove(l[x])
            
        self._sel_last_iid = id
    
    def bind(self,config,fun):
        self.tree.bind(config,fun)
        
    def deleteAll(self):
        for item in self.tree.get_children(''):
            self.tree.delete(item)
        self._items = []
        
    def curselection(self):
        cursel = []
        itmsel = self.tree.selection()
        for i in range(len(self._items)):
            if self._items[i] in itmsel:
                cursel.append(i)
        return cursel
                
    def invertSelection(self):
        for item in self.tree.get_children(''):
            self.tree.selection_toggle(item)
            
    def selectAll(self):
        for item in self.tree.get_children(''):
            self.tree.selection_add(item)
            
    def focus_set(self):
        self.tree.focus_set()

pyLib/nus3bank.py
# coding: utf-8
from __future__ import absolute_import, division, print_function
import pyLib.six as six
import sys
from pyLib.util import *
import os
import string
import zlib

class strg(object):
	name = b""
	size = 0
	offset = 0
	def __init__(self, id):
		self.id = id

class nus3bank(object):
	contents = []
	propOffset = 0
	binfOffset = 0
	grpOffset = 0
	dtonOffset = 0
	toneOffset = 0
	junkOffset = 0
	packOffset = 0
	def __init__(self, size):
		self.size = size

class tone(object):
	name = b""
	ext = b""
	packOffset = 0
	size = 0
	def __init__(self, offset, metaSize):
		self.offset = offset
		self.metaSize = metaSize

def extract(nus3Path,outfolder):
    nus3 = open(nus3Path, u'rb+')
    if nus3.read(4) != b"NUS3":
        nus3.seek(0)
        data = nus3.read()
        nus3_decom = zlib.decompress(data)
        nus3.close()
        nus3 = open(string.replace(nus3Path, u".nus3bank", u"_decompressed.nus3bank"), u"wb+")
        nus3.write(nus3_decom)
        nus3.seek(0)
        data = None
        nus3_decom = None
    else:
        nus3.seek(0)
    
    assert nus3.read(4) == b"NUS3"
    size = readu32le(nus3)

    bank = nus3bank(size)
    assert nus3.read(8) == b"BANKTOC ", u"Not a bank archive!"
    tocSize = readu32le(nus3)
    contentCount = readu32le(nus3)
    offset = 0x14 + tocSize
    for i in range(contentCount):
        content = nus3.read(4)
        contentSize = readu32le(nus3)
        if content == b"PROP":
            propOffset = offset
            propSize = contentSize
        elif content == b"BINF":
            binfOffset = offset
            binfSize = contentSize
        elif content == b"GRP b":
            grpOffset = offset
            grpSize = contentSize
        elif content == b"DTON":
            dtonOffset = offset
            dtonSize = contentSize
        elif content == b"TONE":
            toneOffset = offset
            toneSize = contentSize
        elif content == b"JUNK":
            junkOffset = offset
            junkSize = contentSize
        elif content == b"MARK":
            markOffset = offset
            markSize = contentSize
        elif content == b"PACK":
            packOffset = offset
            packSize = contentSize
        else:
            print(u"Unknown content type " + six.text_type(content))
        offset += 8 + contentSize

    nus3.seek(binfOffset)
    assert nus3.read(4) == b"BINF"
    assert readu32le(nus3) == binfSize
    assert readu32le(nus3) == 0
    assert readu32le(nus3) == 3
    binfStringSize = readByte(nus3)
    binfString = nus3.read(binfStringSize-1)
    # if len(sys.argv) < 3:
        # outfolder = binfString
    # else:
        # outfolder = sys.argv[2]
    if not os.path.exists(outfolder):
        os.mkdir(outfolder)
    # playlistPath = os.path.join(outfolder,'playlist.m3u')
    tracks = []
    nus3.seek(1,1)
    #print binfString
    padding = (binfStringSize + 1) % 4
    #print padding
    if padding == 0:
        pass
    else:
        nus3.seek(abs(padding-4), 1)
    nus3ID = readu32le(nus3)
    #print hex(nus3ID)

    nus3.seek(toneOffset)
    assert nus3.read(4) == b"TONE"
    assert readu32le(nus3) == toneSize
    toneCount = readu32le(nus3)
    print(u"%s:%s:\"%s\"" % (nus3ID, hex(nus3ID), binfString))
    tones = []
    for i in range(toneCount):
        offset = readu32le(nus3) + toneOffset + 8
        metaSize = readu32le(nus3)
        tones.append(tone(offset, metaSize))
	
    for i in range(toneCount):
        if tones[i].metaSize <= 0xc:
            continue
        nus3.seek(tones[i].offset+6)
        tempByte = readByte(nus3)
        if tempByte > 9 or tempByte == 0:
            nus3.seek(5, 1)
        else:
            nus3.seek(1,1)
        stringSize = readByte(nus3)
    #	print hex(nus3.tell())
        tones[i].name = nus3.read(stringSize - 1).decode('utf-8')
        nus3.seek(1,1)
        print(u"\t" + hex(i) + u":" + tones[i].name)
        padding = (stringSize + 1) % 4
        if padding == 0:
            nus3.seek(4, 1)
        else:
            nus3.seek(abs(padding-4) + 4, 1)
    #	assert readu32le(nus3) == 8
        nus3.seek(4,1)
        tones[i].packOffset = readu32le(nus3)
        tones[i].size = readu32le(nus3)
        print((tones[i].packOffset))
    #	print hex(tones[i].packOffset) + b" - b" + hex(tones[i].size)
        if tones[i].packOffset < 0xffffffff:
            nus3.seek(packOffset + 8 + tones[i].packOffset)
            tones[i].ext = u".idsp"

            track = os.path.join(outfolder, hex(i) + u"-" + tones[i].name + tones[i].ext)
            
            # outSound = open(outfolder + b"/" + hex(i) + b"-" + tones[i].name + tones[i].ext, b"wb")
            outSound = open(track,u"wb")
            outSound.write(nus3.read(tones[i].size))
            outSound.close()
            tracks.append(track)
            
            # playlist=open(playlistPath,"a")
            # playlist.write(hex(i) + b"-" + tones[i].name + tones[i].ext+"\n")
            # playlist.close()
        
    nus3.close()
    return tracks